src/Data/TypeRepMap/Internal.hs

```diff
@@ -390,14 +387,19 @@ fromTriples kvs = TypeRepMap (GHC.fromList fpAs) (GHC.fromList fpBs) (GHC.fromLi
 ----------------------------------------------------------------------------
 
 fromSortedList :: forall a . [a] -> [a]
-fromSortedList l = IM.elems $ fst $ go 0 0 mempty (IM.fromList $ zip [0..] l)
+fromSortedList l = runST $ do
+    let n = length l
+    let arrOrigin = fromList l
+    arrResult <- thawArray arrOrigin 0 n
+    _ <- go n 0 0 arrResult arrOrigin
+    toList <$> freezeArray arrResult 0 n
   where
     -- state monad could be used here, but it's another dependency
-    go :: Int -> Int -> IntMap a -> IntMap a -> (IntMap a, Int)
-    go i first result vector =
-      if i >= IM.size vector
-      then (result, first)
+    go :: Int -> Int -> Int -> MutableArray s a -> Array a -> ST s Int
+    go n i first result vector =
```

Comment:
I see that arguments `n`, `result` and `vector` are always passed to `go` function without changes. So it makes sense to me to rearrange and rename argument and introduce one more `where` function to make implementation more clean and robust and efficient:

```haskell
go len result origin = loop  -- or destination/source (dst/src) instead of `result/vector`
  where
    loop = ...
```